rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java

```diff
@@ -441,30 +461,111 @@ BtcBlockStoreWithCache getBtcBlockStore() {
         return btcBlockStore;
     }
 
-    protected PegTxType getTransactionType(BtcTransaction btcTx) {
+    private PeginEvaluationResult evaluatePegin(BtcTransaction btcTx, PeginInformation peginInformation) {
+        if(!activations.isActive(ConsensusRule.RSKIP379)){
+            throw new IllegalStateException("Method cannot be called before RSKIP379.");
+        }
+
         Coin minimumPeginTxValue = bridgeConstants.getMinimumPeginTxValue(activations);
-        Script retiredFederationP2SHScript = provider.getLastRetiredFederationP2SHScript().orElse(null);
 
-        Address oldFederationAddress = Address.fromBase58(
-            bridgeConstants.getBtcParams(),
-            bridgeConstants.getOldFederationAddress()
-        );
+        if(!PegUtils.allUTXOsToFedAreAboveMinimumPeginValue(minimumPeginTxValue, btcTx, getNoSpendWalletForLiveFederations(false), activations)){
+            return new PeginEvaluationResult(PeginProcessAction.CANNOT_BE_PROCESSED, NO_UTXO_OR_UTXO_BELOW_MINIMUM);
+        }
 
-        Wallet federationWallet = new BridgeBtcWallet(btcContext, getLiveFederations());
+        try {
+            peginInformation.parse(btcTx);
+        } catch (PeginInstructionsException e) {
+            String message = String.format(
+                "Error while trying to parse peg-in information for tx %s. %s",
+                btcTx.getHash(),
+                e.getMessage()
+            );
+            logger.warn("[evaluatePegin] {}", message);
+            if (peginInformation.getBtcRefundAddress() != null){
+                return new PeginEvaluationResult(PeginProcessAction.CAN_BE_REFUNDED, PEGIN_V1_INVALID_PAYLOAD);
+            } else {
+                return new PeginEvaluationResult(PeginProcessAction.CANNOT_BE_PROCESSED, PEGIN_V1_INVALID_PAYLOAD);
+            }
+        }
 
-        return PegUtilsLegacy.getTransactionType(
-            btcTx,
-            getActiveFederation(),
-            getRetiringFederation(),
-            retiredFederationP2SHScript,
-            oldFederationAddress,
-            activations,
-            minimumPeginTxValue,
-            federationWallet
+        if (peginInformation.getProtocolVersion() == 0) {
+            switch (peginInformation.getSenderBtcAddressType()) {
+                case P2PKH:
+                case P2SHP2WPKH:
+                    return new PeginEvaluationResult(PeginProcessAction.CAN_BE_REGISTERED);
+                case P2SHMULTISIG:
+                case P2SHP2WSH:
+                    return new PeginEvaluationResult(PeginProcessAction.CAN_BE_REFUNDED, LEGACY_PEGIN_MULTISIG_SENDER);
+                default:
+                    return new PeginEvaluationResult(PeginProcessAction.CANNOT_BE_PROCESSED, LEGACY_PEGIN_UNDETERMINED_SENDER);
+            }
+        } else if (peginInformation.getProtocolVersion() == 1) {
+            return new PeginEvaluationResult(PeginProcessAction.CAN_BE_REGISTERED);
+        }
```

Comment:
I'd move this on top like this
```suggestion
        if (peginInformation.getProtocolVersion() == 1) {
            return new PeginEvaluationResult(PeginProcessAction.CAN_BE_REGISTERED);
        }
        if (peginInformation.getProtocolVersion() == 0) {
        ....
```
Less `else` and possible paths in the code.